### PR TITLE
Clean up FILE_COMMON_ARGUMENTS

### LINKED ISSUE #64956: Modules using add_file_common_args=True and the files document fragment have multiple undocumented arguments
##### SUMMARY
If a module sets `add_file_common_args=True` when calling `AnsibleModule`, all elements from [FILE_COMMON_ARGUMENTS](https://github.com/ansible/ansible/blob/e9d29b1fe4285d90d7a4506b80260a9e24c3bcea/lib/ansible/module_utils/basic.py#L229-L257) are included in the argument spec. The [files document fragment](https://github.com/ansible/ansible/blob/e9d29b1fe4285d90d7a4506b80260a9e24c3bcea/lib/ansible/plugins/doc_fragments/files.py#L15-L80) only documents a subset of them, though. Missing are:
 - src
 - follow
 - force
 - content
 - backup
 - remote_src
 - regexp
 - delimiter
 - directory_mode

Most module authors using `add_file_common_args=True` are probably not aware that their modules have these options as well.

I don't think these extra options should be added by default if `add_file_common_args=True` is specified.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
lib/ansible/module_utils/basic.py

##### ANSIBLE VERSION
```
devel
```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Removes all the options from `FILE_COMMON_ARGUMENTS` which are only there because some action plugins are lazy.

Fixes #64956.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/basic.py
affected (action) modules

### CODE REVIEW

**Comment by felixfontein** [comment]
I think this PR is as ready as it can get (with only my own input). The changes in here potentially break modules and plugins. While I have tried to avoid that, and fix everything I could find, it could still be that there are modules and plugins in ansible/ansible (or somewhere else) which rely on these "invisible" module options being present. I guess the best way to find out is to merge this and see if something breaks, and hopefully find out and fix it before 2.10 is released :)

Once this is merged, I want to update ansible-test (validate modules code) so that it handles the file options as every other option. Right now, the file options are simply ignored, which hides all discrepancies (like the ones mentioned here: https://github.com/ansible/ansible/issues/64956#issuecomment-554854713) and potentially some errors.

**Comment by turb** [comment]
On `maven_artifact`, does the added comment mean that this module should not have this feature, leaving it to others?

**Comment by felixfontein** [comment]
> On `maven_artifact`, does the added comment mean that this module should not have this feature, leaving it to others?

It means that it is totally unclear to me whether it should have that feature or not. The module never explicitly declared that option or documented it, but apparently @laurentribot was aware of it, since his PR uses it. Though that might have just been because parts of the code in that PR were taken from the `copy` module (https://github.com/ansible/ansible/pull/24965#issuecomment-320927623).

Anyway, I wanted to avoid random breakage by removing the option, so I re-added it manually.

File: changelogs/fragments/66389-file-common-arguments.yml
Comment (jillr) [inline changelogs/fragments/66389-file-common-arguments.yml]:
```suggestion
   ``delimiter``, and ``directory_mode`` are now no longer added. Modules relying on these options
```

File: docs/docsite/rst/porting_guides/porting_guide_2.10.rst
Comment (jillr) [inline docs/docsite/rst/porting_guides/porting_guide_2.10.rst]:
```suggestion
* Ansible modules created with ``add_file_common_args=True`` added a number of undocumented arguments which were mostly there to ease implementing certain action plugins. The undocumented arguments ``src``, ``follow``, ``force``, ``content``, ``backup``, ``remote_src``, ``regexp``, ``delimiter``, and ``directory_mode`` are now no longer added. Modules relying on these options to be added need to specify them by themselves.
```

**Review by jillr — COMMENTED** [review]
Couple of minor nits

**Comment by jillr** [comment]
@felixfontein as discussed in irc meeting today, please ci_complete and run full tests a few times to try to flush out any possible breakage.

**Comment by felixfontein** [comment]
First successful run: https://app.shippable.com/github/ansible/ansible/runs/157640/summary/console

**Comment by felixfontein** [comment]
Another successful run: https://app.shippable.com/github/ansible/ansible/runs/157727/summary/console

(There would be more, but sometimes after several AWS retries I get "please rerun the entire matrix" before it completes...)

**Comment by felixfontein** [comment]
Third successful run: https://app.shippable.com/github/ansible/ansible/runs/157758/summary/console

**Comment by felixfontein** [comment]
Forth successful run: https://app.shippable.com/github/ansible/ansible/runs/157799/summary/console

**Review by jillr — APPROVED** [review]

File: docs/docsite/rst/porting_guides/porting_guide_2.10.rst
Comment (samccann) [inline docs/docsite/rst/porting_guides/porting_guide_2.10.rst:95]: Is there any way a module owner would know they used that ``add_file_common_args=True`` ?  And what happens to a user of one of those modules if the module owner doesn't make a code change in 2.10 to match this?

Comment (felixfontein) [reply docs/docsite/rst/porting_guides/porting_guide_2.10.rst:95]: Module owners explicitly have to set `add_file_common_args=True`, so they should know. For most modules using `add_file_common_args=True`, this should have absolutely no effect, since these parameters weren't documented and the sanity checks ignore them explicitly (i.e. essentially nobody knows that they are undocumented, and thus that they exist at all). It can happen in a few cases (like the ones I fixed in this PR) that modules forgot to declare some of these options by themselves, because "it just worked".

There are 29 modules in ansible/ansible which use this flag. These ones we can check manually (a lot of work, but not impossible), but if someone has a custom module which depends on one of these options, we can't really avoid this. But for this similar arguments as for the modules in ansible/ansible apply: this shouldn't affect most modules using `add_file_common_args=True`, and a huge majority of modules doesn't even use that.

Comment (samccann) [reply docs/docsite/rst/porting_guides/porting_guide_2.10.rst:95]: ok thanks for the explanation!

**Review by samccann — APPROVED** [review]
docs portion LGTM

**Comment by felixfontein** [comment]
I went manually through all 29 modules in ansible/ansible having `add_file_common_args=True`, all of them seem to delcare the parameters themselves (or don't use them). (Note that `force` is also defined by `url_argument_spec` in `module_utils/urls.py`, and the `get_url` and `uri` modules get that param from there.)

**Review by Andersson007 — COMMENTED** [review]
in part of postgresql LGTM.
In general too, but we need to be sure all cases were handled. Maybe some of the modules don't have CI tests (or proper CI tests), etc.

**Comment by Andersson007** [comment]
if i were you i'd grep through CI tests for those modules which use ```add_file_common_args=True``` to be sure the parameters are tested.

**Review by resmo — APPROVED** [review]
LGTM

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
Thanks @felixfontein for cleaning this up and everyone who reviewed it!

**Comment by felixfontein** [comment]
@s-hertel also thanks a lot for reviewing and merging! And of course to everyone else for reviewing! :)

**Pull request merged**
